lib/ansible/plugins/action/__init__.py

```diff
@@ -242,18 +242,22 @@ def _configure_module(self, module_name, module_args, task_vars=None):
                 # we'll propagate back to the controller in the task result
                 discovered_key = 'discovered_interpreter_%s' % idre.interpreter_name
 
+                # update the local vars copy for the retry
+                if use_vars.get('ansible_facts') is None:
+                    use_vars['ansible_facts'] = {}
+                use_vars['ansible_facts'][discovered_key] = self._discovered_interpreter
+
                 # TODO: this condition prevents 'wrong host' from being updated
                 # but in future we would want to be able to update 'delegated host facts'
                 # irrespective of task settings
                 if not self._task.delegate_to or self._task.delegate_facts:
                     # store in local task_vars facts collection for the retry and any other usages in this worker
-                    if use_vars.get('ansible_facts') is None:
+                    if task_vars.get('ansible_facts') is None:
```

Comment:
we should always have ansible_facts here, the if/set to empty should not be needed here, nor above.